### DISCUSSION TITLE
How can I use Prodigy Optimiser?

### DISCUSSION THREAD

**Discussion by Reviem0** [Linaqruf/kohya-trainer#273] [Q&A]
I'd like to use Prodigyopt with the kohya Lora trainer colab, how can I add it in?

**Comment by Eden-Luminar** [comment]
Prodigy Optimiser is desingned to be used to train?

**Comment by Reviem0** [comment]
Thanks @MetroByte for the answer

> If this is relevant for you: For "kohya-trainer" google colab:
> 
> 1. Add a code block (on top, for example) with code `!pip install -U prodigyopt`
> 2. Proceed setup as usual up to 5.3
> 3. You have to change your settings a bit in `5.3`-`5.5`:
> 
> * In `optimizer_args` set the following _(you can change these if you know what are you doing)_: `["decouple=True", "weight_decay=0.01", "d_coef=2", "use_bias_correction=True", "safeguard_warmup=True"]`
> * `unet_lr`, `text_encoder_lr` to **1 (important)**
> 
> 5. Then you have to change your config files in folder `LoRA/config`. In `config_file.toml`:
> 
> * Make sure that every `lr` (learning rates) are set to _(probably)_ **0.1 to 1, or not set at all (otherwise error, needs checking)** [Error when trying to train using DAdaptation optimizer bmaltais/kohya_ss#1160 (comment)](https://github.com/bmaltais/kohya_ss/issues/1160#issuecomment-1631606123)
> * `optimizer_args` is present
> * `optimizer_type` is `Prodigy` (safecheck)
> 
> **Then you have to change a bit in the `kohya-trainer/train_network.py`**
> 
> 1. At line ~12-13 add `from prodigyopt import Prodigy`
> 2. Find and comment line _(with `#`)_ containing text `optimizer_name, optimizer_args, optimizer = …` (Line number ~224-227)
> 3. Add the following three lines below the commented line:
> 
> `optimizer_name = "Prodigy"` `optimizer_args = "['decouple': True, 'weight_decay': 0.01, 'd_coef': 2, 'use_bias_correction': True, 'safeguard_warmup': True]"` `optimizer = Prodigy(trainable_params)`
> 
> _(first two are for metadata printing only though)_
> 
> Then training should run fine (I hope)
> 
> If you re-run steps 5.2-5.4, config file will be replaced, so better to change config files directly.
> 
> _Tested but I get overfitting results so early, I don't know how to set Prodigy correctly…_
